### PR TITLE
Chamel/equinox functionspace

### PR DESCRIPTION
**Pull request by cmhamel**

### CODE REVIEW

**Comment by cmhamel** [comment]
Got some mild start up improvements after I found some jax.numpy <-> numpy conversions in the DofManager setup.

File: optimism/FunctionSpace.py [hunk 1]
```diff
@@ -354,25 +363,43 @@ def integrate_function_on_edges(functionSpace, func, U, quadRule, edges):
     return np.sum(integrate_on_edges(functionSpace, func, U, quadRule, edges))
 
 
-class DofManager:
+class DofManager(eqx.Module):
+    # TODO get type hints below correct
+    # TODO this one could be moved to jax types if we move towards
+    # TODO jit safe preconditioners/solvers
```
Comment (ralberd) [inline optimism/FunctionSpace.py:369]: @btalamini Is this something you can help with?

File: optimism/FunctionSpace.py [hunk 2]
```diff
@@ -354,25 +363,43 @@ def integrate_function_on_edges(functionSpace, func, U, quadRule, edges):
...
-        self.ids = np.arange(self.isBc.size).reshape(self.fieldShape)
+        self.ids = onp.arange(self.isBc.size).reshape(self.fieldShape)
 
         self.unknownIndices = self.ids[self.isUnknown]
         self.bcIndices = self.ids[self.isBc]
 
-        ones = np.ones(self.isBc.size, dtype=int) * -1
-        self.dofToUnknown = ones.at[self.unknownIndices].set(np.arange(self.unknownIndices.size)) 
+        ones = onp.ones(self.isBc.size, dtype=int) * -1
+        # self.dofToUnknown = ones.at[self.unknownIndices].set(np.arange(self.unknownIndices.size))
```
Comment (ralberd) [inline optimism/FunctionSpace.py:395]: Can you remove the commented code?

Comment (cmhamel) [reply optimism/FunctionSpace.py]: Done

File: optimism/FunctionSpace.py (hunk 1 above)
Comment (cmhamel) [reply optimism/FunctionSpace.py:369]: I can circle back for these type hints later on. See this for some canned solvers [lineax](https://github.com/patrick-kidger/lineax)

**Review by ralberd — APPROVED** [review]

**Pull request merged**